pkg/controller/job/job_controller.go

```diff
@@ -861,7 +864,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 		}
 	}
 
-	if isIndexedJob(&job) {
+	if !isSuccessCriteriaMetCondition(jobCtx.finishedCondition) && isIndexedJob(&job) {
```

Comment:
Why is this check needed? Maybe leave a comment.

Reply:
This is needed to avoid the transitions of SuccessCriteriaMet -> Failed.
So, I'll try to explain the reason here.

Reply:
but this transition wouldn't happen because setting the `jobCtx.finishedCondition ` is already gaurded by `jobCtx.finishedCondition == nil` check inside, no?

Reply:
@mimowo Yes, that's right, but shouldn't we avoid to be updated the `prevSucceededIndexes`,`succeeded`, and `failedIndexes`?

Reply:
If there are some pods which completed successfully in the meanwhile I think it is desired to account for them. 

Similarly if there are some indexes that failed in the meanwhile I think it is desired to update the set of failed indexes computed for backoffLimitPerIndex `jobCtx.failedIndexes = calculateFailedIndexes(logger, &job, pods)` . 

Maybe you are concerned that with the updates we no longer satisfy the criteria? I'm not sure this is possible, but a similar thing could happen for pod failuire policy. We could have a pod which triggered FailJob, so we set FailureTarget to remember the decision. Because it is possible that the controller crashes, and the pod is garbage collected, then we would lose the signal FailJob happened. Here also, we record the fact the criteria were met at some point in time.

I think as long as we guarantee that SucceessCriteriaMet transitions to Complete, and FailureTarget to Failed, we are good.

Reply:
> Because it is possible that the controller crashes, and the pod is garbage collected, then we would lose the signal FailJob happened. Here also, we record the fact the criteria were met at some point in time.

I guess that finalizers would block the removal of non-recorded pods. So, I believe that it could not be happened.

@mimowo Anyway, since my concerns have gone away, I'm ok with removing `!isSuccessCriteriaMetCondition(jobCtx.finishedCondition) &&`. Thank you for clarifications!

Reply:
> I guess that finalizers would block the removal of non-recorded pods. So, I believe that it could not be happened.

Right, it would be possible if we kept the finalizer for pods matching FailJob, but thanks to recording the FailureTarget condition we can remove their finalizers as soon as possible (as for any other failed pods).

Reply:
Done.